ovos_core/skill_installer.py

```diff
@@ -0,0 +1,174 @@
+from tempfile import gettempdir
+from typing import Optional
+
+from combo_lock import ComboLock
+from ovos_config.config import Configuration
+
+from ovos_bus_client import Message
+from ovos_utils.log import LOG
+from importlib import reload
+import ovos_plugin_manager
+
+
+class SkillsStore:
+    # default constraints to use if none are given
+    DEFAULT_CONSTRAINTS = '/etc/mycroft/constraints.txt'
+    PIP_LOCK = ComboLock(join(gettempdir(), "ovos_pip.lock"))
+
+    def __init__(self, bus, config=None):
+        self.config = config or Configuration().get("skills", {}).get("installer", {})
+        self.bus = bus
+        self.bus.on("ovos.skills.install", self.handle_install_skill)
+        self.bus.on("ovos.skills.uninstall", self.handle_uninstall_skill)
+        self.bus.on("ovos.pip.install", self.handle_install_python)
+        self.bus.on("ovos.pip.uninstall", self.handle_uninstall_python)
+
+    def shutdown(self):
+        pass
+
+    def play_error_sound(self):
+        self.bus.emit(Message("mycroft.audio.play_sound",
+                              {"uri": "snd/error.mp3"}))
+
+    def play_success_sound(self):
+        self.bus.emit(Message("mycroft.audio.play_sound",
+                              {"uri": "snd/acknowledge.mp3"}))
+
+    def pip_install(self, packages: list,
+                    constraints: Optional[str] = None,
+                    print_logs: bool = True):
+        if not len(packages):
+            self.play_error_sound()
+            return False
+        # Use constraints to limit the installed versions
+        if constraints and not exists(constraints):
+            LOG.error('Couldn\'t find the constraints file')
+            self.play_error_sound()
+            return False
+        elif exists(SkillsStore.DEFAULT_CONSTRAINTS):
+            constraints = SkillsStore.DEFAULT_CONSTRAINTS
+
+        pip_args = [sys.executable, '-m', 'pip', 'install']
+        if constraints:
+            pip_args += ['-c', constraints]
+        if self.config.get("break_system_packages", False):
+            pip_args += ["--break-system-packages"]
+
+        with SkillsStore.PIP_LOCK:
+            """
+            Iterate over the individual Python packages and
+            install them one by one to enforce the order specified
+            in the manifest.
+            """
+            for dependent_python_package in packages:
+                LOG.info("(pip) Installing " + dependent_python_package)
+                pip_command = pip_args + [dependent_python_package]
+                LOG.debug(" ".join(pip_command))
+                if print_logs:
+                    proc = Popen(pip_command)
+                else:
+                    proc = Popen(pip_command, stdout=PIPE, stderr=PIPE)
+                pip_code = proc.wait()
+                if pip_code != 0:
+                    stderr = proc.stderr.read().decode()
+                    self.play_error_sound()
+                    raise RuntimeError(stderr)
```

Comment:
Consider a config flag to run the pip install for everything at once, to increase performance.

Reply:
this was a conscious decision to follow specified package order, eg, think wheel, swig, cython needing to be installed first or the other packages fail

logic comes from OSM, we can revisit at a later point or PRs welcome to make configurable